ts4.1/index.d.ts

```diff
@@ -0,0 +1,294 @@
+	 *
+	 * It simulates `camelcase`'s behavior. Therefore.
+	 *
+	 * - Underscores `_`, hyphens `-`, periods `.`, and whitespaces ` ` are treated as word separators.
+	 * - Leading and trailing separators are removed during preprocessing.
+	 * - Consecutive separators work the same as one separator.
+	 *
+	 * @param S a union of string literal types to camelcase.
+	 * @param PascalCase if true, the first word is also capitalized.
+	 *
+	 * @example
+	 * ```
+	 * type T1 = Camelize<"foo_bar-baz">;
+	 * // => "fooBarBaz"
+	 *
+	 * type T2 = Camelize<"__foo_bar__" | "some type">;
+	 * // => "fooBar" | "someType"
+	 *
+	 * type T3 = Camelize<string>;
+	 * // => string
+	 *
+	 * type T4 = Camelize<"foo_bar", true>;
+	 * // => "FooBar"
+	 *
+	 * type T5 = Camelize<"foo_bar", boolean>;
+	 * // => "fooBar" | "FooBar"
+	 * ```
+	 */
+	type Camelize<S extends string, PascalCase extends boolean = false> =
+		S extends `_${infer S2}` ? Camelize<S2, PascalCase> :
+		S extends `-${infer S2}` ? Camelize<S2, PascalCase> :
+		S extends `.${infer S2}` ? Camelize<S2, PascalCase> :
+		S extends ` ${infer S2}` ? Camelize<S2, PascalCase> :
+		S extends `${infer S1}_${infer S2}` ? `${Camelize<S1, PascalCase>}${Camelize<S2, true>}` :
+		S extends `${infer S1}-${infer S2}` ? `${Camelize<S1, PascalCase>}${Camelize<S2, true>}` :
+		S extends `${infer S1}.${infer S2}` ? `${Camelize<S1, PascalCase>}${Camelize<S2, true>}` :
+		S extends `${infer S1} ${infer S2}` ? `${Camelize<S1, PascalCase>}${Camelize<S2, true>}` :
+		PascalCase extends true ? Capitalize<S> : S;
+
+	/**
+	 * @internal
+	 * Digs a set of paths by one depth. Used for processing `stopPaths`.
+	 *
+	 * @param Path a union of string literal types, each of which is a "path" from the root.
+	 * @param S a prefix word. A string literal type or `string`.
+	 *
+	 * @example
+	 * ```
+	 * type T1 = FilterPaths<"foo.bar" | "foo.baz" | "foo.bar.baz" | "other.path", "foo">;
+	 * // => "bar" | "baz" | "bar.baz"
+	 *
+	 * type T2 = FilterPaths<string, "foo">;
+	 * // => string
+	 *
+	 * type T3 = FilterPaths<"foo.bar" | "foo.baz" | "foo.bar.baz" | "other.path", string>;
+	 * // => string
+	 *
+	 * type T3 = FilterPaths<"foo" | "bar", string>;
+	 * // => never
+	 * ```
+	 */
+	type FilterPaths<Path extends string, S extends string> =
+		// We treat `string` as "indeterminate set of strings", so return `string` in this case.
+		string extends Path ? string :
+		// If `S` is indeterminate, we are not sure about the resulting set,
+		// except when there is no path containing `.`.
+		string extends S ?
+		Path extends `${string}.${string}` ? string : never :
+		// Otherwise, do a normal filtering.
+		// If a path starts with `${S}.`, then return the suffix.
+		// Otherwise return never (which means the candidate is rejected)
+		Path extends `${S}.${infer Suffix}` ? Suffix : never;
+
+	/**
+	 * @internal
+	 * Helper type to conditionally camelize a string.
+	 *
+	 * @param S a union of string literal types to camelcase.
```

Comment:
The other type variables are descriptive, so I think this one should be too.

Reply:
Unfortunately, I could only think of `AStringToBeCamelized` and none of `A`, `String`, `To`, `Be`, and `Camelized` seemed necessary to me.